File: StudyGamePrograming08/VertexInfo.cpp
#include "VertexInfo.h"
#include <glew.h>

VertexInfo::VertexInfo(const float* vertPos, const float* texCoord, 
					   const float* vertColor, const unsigned int* indices,
					   unsigned int numVerts, unsigned int numIndices)
	: mNumVerts(numVerts), mNumIndices(numIndices)
{
	// バーテックス配列オブジェクトをOpenGLに生成し、そのIDをメンバー変数mVertexArrayに保存する
	glGenVertexArrays(1, &mVertexArray);
	glBindVertexArray(mVertexArray);

	// インデックスバッファをOpenGLに生成し、そのIDをメンバー変数mIndexBufferに保存する
	glGenBuffers(1, &mIndexBuffer);
	glBindBuffer(GL_ELEMENT_ARRAY_BUFFER, mIndexBuffer);
	glBufferData(
		GL_ELEMENT_ARRAY_BUFFER,			//インデックスバッファの指定
		mNumIndices * sizeof(unsigned int),	//データのサイズ
		indices,							//コピー元の配列
		GL_STATIC_DRAW						//このデータを1回だけロードしてその後頻繁に使うとき
	);

	//VertexAttribute layout0 = position
	unsigned int cnt = 3;					//要素数
	glGenBuffers(1, &mVertPosBuffer);
	glBindBuffer(GL_ARRAY_BUFFER, mVertPosBuffer);
	glBufferData(
		GL_ARRAY_BUFFER,					// バッファの種類にバーテックスバッファを指定
		cnt * mNumVerts * sizeof(float),	//コピーするバイト数
		vertPos,							//コピー元の配列
		GL_STATIC_DRAW						//このデータを1回だけロードしてその後頻繁に使うとき
	);
	glVertexAttribPointer(
		0,									// 属性インデックス（１つ目はゼロ）
		cnt,								// 要素数
		GL_FLOAT,							// 要素の型
		GL_FALSE,							// 整数型のみ使用する。
		cnt * sizeof(float),				// ストライド（通常は各バーテックス属性のデータ数
		0									// 頂点データの開始位置からこの属性までのオフセット
	);
	glEnableVertexAttribArray(0);

	// VertexAttribute layout1 = texCoord
	cnt = 2;		//テクスチャ座標配列の要素数
	glGenBuffers(1, &mTexCoordBuffer);
	glBindBuffer(GL_ARRAY_BUFFER, mTexCoordBuffer);
	glBufferData(
		GL_ARRAY_BUFFER,					// バッファの種類にバーテックスバッファを指定
		cnt * mNumVerts * sizeof(float),	//コピーするバイト数
		texCoord,							//コピー元の配列
		GL_STATIC_DRAW						//このデータを1回だけロードしてその後頻繁に使うとき
	);
	glVertexAttribPointer(
		1,									// 属性インデックス（１つ目はゼロ）
		cnt,								// 要素数
		GL_FLOAT,							// 要素の型
		GL_FALSE,							// 整数型のみ使用する。
		cnt * sizeof(float),				// ストライド（通常は各バーテックス属性のデータ数
		0									// 頂点データの開始位置からこの属性までのオフセット
	);
	glEnableVertexAttribArray(1);

	// VertexAttribute layout2 = vertColor
	cnt = 4;								//頂点カラー配列の要素数
	glGenBuffers(1, &mVertColorBuffer);
	glBindBuffer(GL_ARRAY_BUFFER, mVertColorBuffer);
	glBufferData(
		GL_ARRAY_BUFFER,					// バッファの種類にバーテックスバッファを指定
		cnt * mNumVerts * sizeof(float),	//コピーするバイト数
		vertColor,							//コピー元の配列
		GL_STATIC_DRAW						//このデータを1回だけロードしてその後頻繁に使うとき
	);
	glVertexAttribPointer(
		2,									// 属性インデックス（１つ目はゼロ）
		cnt,								// 要素数
		GL_FLOAT,							// 要素の型
		GL_FALSE,							// 整数型のみ使用する。
		cnt * sizeof(float),				// ストライド（通常は各バーテックス属性のデータ数
		0									// 頂点データの開始位置からこの属性までのオフセット
	);
	glEnableVertexAttribArray(2);

}


VertexInfo::~VertexInfo()
{
	glDeleteBuffers(1, &mVertPosBuffer);
	glDeleteBuffers(1, &mIndexBuffer);
	glDeleteBuffers(1, &mTexCoordBuffer);
	glDeleteBuffers(1, &mVertColorBuffer);
	glDeleteVertexArrays(1, &mVertexArray);
}

void VertexInfo::SetActive()
{
	glBindVertexArray(mVertexArray);
}

File: StudyGamePrograming08_VB/VertexInfo.vb
Imports System.Runtime.InteropServices
Imports OpenTK
Imports OpenTK.Graphics
Imports OpenTK.Graphics.OpenGL

Public Class VertexInfo
    Implements IDisposable      '明示的にクラスを開放するために必要

    'public
    Public Sub New(ByRef vertPos As Single(), ByRef texCoord As Single(),
                   ByRef vertColor As Single(), ByRef indices As UInteger(),
                   ByVal numVerts As Integer, ByVal numIndices As Integer)
        mNumVerts = numVerts
        mNumIndices = numIndices

        mVertexArray = GL.GenVertexArray()
        GL.BindVertexArray(mVertexArray)

        mIndexBuffer = GL.GenBuffer()
        GL.BindBuffer(BufferTarget.ElementArrayBuffer, mIndexBuffer)
        GL.BufferData(BufferTarget.ElementArrayBuffer,
                      mNumIndices * Marshal.SizeOf(Of UInteger),
                      indices,
                      BufferUsageHint.StaticDraw)

        'VertexAttribute layout0 = position
        Dim cnt As Integer = 2      '要素数
        mVertPosBuffer = GL.GenBuffer()
        GL.BindBuffer(BufferTarget.ArrayBuffer, mVertPosBuffer)
        GL.BufferData(BufferTarget.ArrayBuffer,
                      cnt * mNumVerts * Marshal.SizeOf(Of Single),
                      vertPos,
                      BufferUsageHint.StaticDraw)
        GL.VertexAttribPointer(0,
                               cnt,
                               VertexAttribPointerType.Float,
                               False,
                               cnt * Marshal.SizeOf(Of Single),
                               0)
        GL.EnableVertexAttribArray(0)

        'VertexAttribute layout1 = texCoord
        cnt = 2
        mTexCoordBuffer = GL.GenBuffer()
        GL.BindBuffer(BufferTarget.ArrayBuffer, mTexCoordBuffer)
        GL.BufferData(BufferTarget.ArrayBuffer,
                      cnt * mNumVerts * Marshal.SizeOf(Of Single),
                      texCoord,
                      BufferUsageHint.StaticDraw)
        GL.VertexAttribPointer(1,
                               cnt,
                               VertexAttribPointerType.Float,
                               False,
                               cnt * Marshal.SizeOf(Of Single),
                               0)
        GL.EnableVertexAttribArray(1)

        'VertexAttribute layout2 = vertColor
        cnt = 4
        mVertColorBuffer = GL.GenBuffer()
        GL.BindBuffer(BufferTarget.ArrayBuffer, mVertColorBuffer)
        GL.BufferData(BufferTarget.ArrayBuffer,
                      cnt * mNumVerts * Marshal.SizeOf(Of Single),
                      vertColor,
                      BufferUsageHint.StaticDraw)
        GL.VertexAttribPointer(2,
                               cnt,
                               VertexAttribPointerType.Float,
                               False,
                               cnt * Marshal.SizeOf(Of Single),
                               0)
        GL.EnableVertexAttribArray(2)

    End Sub
    Protected disposed = False     '開放処理が実施済みかのフラグ
    Public Overloads Sub Dispose() Implements IDisposable.Dispose
        Dispose(True)
    End Sub
    Protected Overridable Overloads Sub Dispose(ByVal disposing As Boolean)
        If Not Me.disposed Then
            If disposing Then
                '*** アンマネージリソースの開放
                GL.DeleteBuffers(1, mVertPosBuffer)
                GL.DeleteBuffers(1, mTexCoordBuffer)
                GL.DeleteBuffers(1, mVertColorBuffer)
                GL.DeleteBuffers(1, mIndexBuffer)
                GL.DeleteVertexArrays(1, mVertexArray)
            End If
            '*** マネージドリソースの開放
        End If
        disposed = True
    End Sub
    Protected Overrides Sub Finalize()
        MyBase.Finalize()
        Dispose(False)
    End Sub

    Public Sub SetActive()
        GL.BindVertexArray(mVertexArray)
    End Sub
    Public Function GetNumVerts() As Integer
        Return mNumVerts
    End Function
    Public Function GetNumIndices() As Integer
        Return mNumIndices
    End Function

    'private
    Private disposedValue As Boolean
    Private mNumVerts As Integer            ' 頂点の数
    Private mNumIndices As Integer          ' インデックスの数
    Private mVertexArray As Integer         ' バーテックス配列オブジェクトのOpenGL ID
    Private mIndexBuffer As Integer         ' インデックスバッファのOpenGL ID
    Private mVertPosBuffer As Integer       ' 頂点座標バッファのOpenGL ID
    Private mTexCoordBuffer As Integer      ' テクスチャ座標バッファのOpenGL ID
    Private mVertColorBuffer As Integer     ' 頂点カラーバッファのOpenGL ID

End Class
